internal/controlplane/common.go

```diff
@@ -59,3 +67,32 @@ func getProjectFromRequestOrDefault(ctx context.Context, in ProjectIDGetter) (uu
 	}
 	return parsedProjectID, nil
 }
+
+func getProviderFromRequestOrDefault(
+	ctx context.Context,
+	store db.Store,
+	in ProviderNameGetter,
+	projectId uuid.UUID,
+) (db.Provider, error) {
+	providers, err := store.ListProvidersByProjectID(ctx, projectId)
+	if err != nil {
+		return db.Provider{}, status.Errorf(codes.InvalidArgument, "cannot retrieve providers: %s", err)
+	}
+	// if we do not have a provider name, check if we can infer it
+	if in.GetProvider() == "" {
+		if len(providers) == 1 {
+			return providers[0], nil
+		}
+		return db.Provider{}, status.Errorf(codes.InvalidArgument, "cannot infer provider name: %s", err)
+	}
+
+	matchesName := func(provider db.Provider) bool {
+		return provider.Name == in.GetProvider()
+	}
+
+	i := slices.IndexFunc(providers, matchesName)
+	if i == -1 {
+		return db.Provider{}, status.Errorf(codes.InvalidArgument, "invalid provider name: %s", err)
```

Comment:
(Again, `err` is `nil` here.  You probably also want to use `util.UserVisibleError`)

```suggestion
		return db.Provider{}, util.UserVisibleError(codes.InvalidArgument, "invalid provider name: %s", in.GetProvider())
```